### PR TITLE
Added new option to disable purging rules when updating ALB

### PR DESCRIPTION
**Pull request by jcortejoso**
##### SUMMARY

Added the parameter _purge_rules_ to _elb_application_lb_ to permit keeping the listener rules if the listener already exists and has previous rules. 

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

elb_application_lb
##### ANSIBLE VERSION

```
ansible 2.4.0.0
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Review by s-hertel — APPROVED** [review]

**Comment by willthames** [comment]
Could do with some integration tests, but as elb_application_lb is preview, and doesn't yet have tests, this change meets current policies.

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py [hunk 1]
```diff
@@ -665,6 +671,7 @@ def compare_rules(connection, module, current_listeners, listener):
     :param module: Ansible module object
     :param current_listeners: list of listeners currently associated with the ELB
     :param listener: dict object of a listener passed by the user
+    :param purge_rules: boolean to indicate if the exisisting rules for the listener must be purged
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: typo: existing

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py [hunk 2]
```diff
@@ -68,6 +68,12 @@
         not set then listeners will not be modified
     default: yes
     choices: [ 'yes', 'no' ]
+  purge_rules:
+    description:
+      - If yes, existing rules will be purged from the ELB to match exactly the I(Rules) defined in each I(listeners) parameter.
+    default: yes
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/elb_application_lb.py]: I'd rather this default to `no`. It is a change, but users are more likely to come a cropper on this being true than false.

**Review by willthames — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py (hunk 2 above)
Comment (jcortejoso) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Ok for me. I tried to keep the same behavior that it had before the PR. @wimnat any thought?

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py (hunk 1 above)
Comment (jcortejoso) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py]: Thanks 👍

**Comment by hugobordigoni** [comment]
@jcortejoso thanks for the feature, it's exactly what I need.
Any idea when it's gonna be merged?

File: lib/ansible/modules/cloud/amazon/elb_application_lb.py (hunk 2 above)
Comment (hugobordigoni) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py]: My 2 cts on this. I think it should have same behaviour as `purge_listeners`, meaning having a default value `yes`.
Though, I'll definitely use this feature and set it to no!

Comment (wimnat) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py]: I always consider ansible to be first and foremost trying to attain desired state which is why I default to yes. It's a mixture across other modules. I'm not too fussed but whatever the decision it should be documented so we can get some consistency

Comment (willthames) [reply lib/ansible/modules/cloud/amazon/elb_application_lb.py]: And I always remember this tweet: https://twitter.com/stigglor/status/888221882360582145

**Comment by jcortejoso** [comment]
I updated the PR with purge_rules = False by default. I would consider setting purge_listeners and purge_tags also False to be more consistent. 
I can look if I am able to add some tests. Is there any guide or documentation related with the tests?

**Comment by willthames** [comment]
@jcortejoso - I wrote a blog post on testing AWS modules: http://willthames.github.io/2017/07/17/so-you-want-to-test-aws-modules-for-ansible.html

The ec2_group test suite is probably the most comprehensive example at the moment.

**Comment by wimnat** [comment]
I'm going to try to include this functionality as part of ALB rewrite or shortly there after.

**Comment by chenl87** [comment]
+1
Awesome feature. Any ETA for this?

**Comment by gundalow** [comment]
Think this has been resolved by https://github.com/ansible/ansible/pull/43113

**Pull request closed without merging**